### ISSUE TITLE
ModelsCommand: array_diff(): Expected parameter 1 to be an array, null given

### ISSUE THREAD

**Issue by mfn** [barryvdh/laravel-ide-helper#1112]
### Versions:
- ide-helper Version: dev-master 9859958
- Laravel Version: 8.16.1
- PHP Version: 7.1.12

### Description:
```
Exception: array_diff(): Expected parameter 1 to be an array, null given
Could not analyze class App\Models\SomeModel.

Trace:
#0 [internal function]: Illuminate\Foundation\Bootstrap\HandleExceptions->handleError(2, 'array_diff(): E...', '/vagrant/api.sw...', 1163, Array)
#1 vendor/barryvdh/laravel-ide-helper/src/Console/ModelsCommand.php(1163): array_diff(NULL, Array)
#2 vendor/barryvdh/laravel-ide-helper/src/Console/ModelsCommand.php(565): Barryvdh\LaravelIdeHelper\Console\ModelsCommand->writeModelExternalBuilderMethods('Builder', Object(App\Models\SomeModel))
#3 vendor/barryvdh/laravel-ide-helper/src/Console/ModelsCommand.php(277): Barryvdh\LaravelIdeHelper\Console\ModelsCommand->getPropertiesFromMethods(Object(App\Models\SomeModel))
#4 vendor/barryvdh/laravel-ide-helper/src/Console/ModelsCommand.php(159): Barryvdh\LaravelIdeHelper\Console\ModelsCommand->generateDocs(Array, Array)
#5 vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(36): Barryvdh\LaravelIdeHelper\Console\ModelsCommand->handle()
#6 vendor/laravel/framework/src/Illuminate/Container/Util.php(40): Illuminate\Container\BoundMethod::Illuminate\Container\{closure}()
#7 vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(93): Illuminate\Container\Util::unwrapIfClosure(Object(Closure))
#8 vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(37): Illuminate\Container\BoundMethod::callBoundMethod(Object(App\Services\Vendor\Illuminate\Foundation\Application), Array, Object(Closure))
#9 vendor/laravel/framework/src/Illuminate/Container/Container.php(610): Illuminate\Container\BoundMethod::call(Object(App\Services\Vendor\Illuminate\Foundation\Application), Array, Array, NULL)
#10 vendor/laravel/framework/src/Illuminate/Console/Command.php(136): Illuminate\Container\Container->call(Array)
#11 vendor/symfony/console/Command/Command.php(258): Illuminate\Console\Command->execute(Object(Symfony\Component\Console\Input\ArgvInput), Object(Illuminate\Console\OutputStyle))
#12 vendor/laravel/framework/src/Illuminate/Console/Command.php(121): Symfony\Component\Console\Command\Command->run(Object(Symfony\Component\Console\Input\ArgvInput), Object(Illuminate\Console\OutputStyle))
#13 vendor/symfony/console/Application.php(920): Illuminate\Console\Command->run(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#14 vendor/symfony/console/Application.php(266): Symfony\Component\Console\Application->doRunCommand(Object(Barryvdh\LaravelIdeHelper\Console\ModelsCommand), Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#15 vendor/symfony/console/Application.php(142): Symfony\Component\Console\Application->doRun(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#16 vendor/laravel/framework/src/Illuminate/Console/Application.php(93): Symfony\Component\Console\Application->run(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#17 vendor/laravel/framework/src/Illuminate/Foundation/Console/Kernel.php(129): Illuminate\Console\Application->run(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#18 artisan(37): Illuminate\Foundation\Console\Kernel->handle(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#19 {main}
```

**Comment by mfn** [comment]
The `$builder` received in https://github.com/barryvdh/laravel-ide-helper/blob/9859958cfaf522b8e6ef13352067f9ef82f9e8b5/src/Console/ModelsCommand.php#L1152
which is supposed to the FQCN, is only `Builder` in my case.

Seems the namespace is not properly provided in some cases.

**Comment by mfn** [comment]
It's also related to the config setting `force_fqn`:
- if it's `true`, it works and I don't get an exception
- with `false`, the default value, it bombs

**Comment by mfn** [comment]
I've to stop now, if someone experiences this too and wants to investigate, feel free!

**Comment by mfn** [comment]
@KaloyanYosifov can you help here?

After some debugging I found out it's related to https://github.com/barryvdh/laravel-ide-helper/pull/1089

Specifically I think there's a problem here https://github.com/barryvdh/laravel-ide-helper/pull/1089/files#diff-2df8ee218526540f71334c651990edfbfe5345cc53a521b0259c874ff24abd9aR1152
```php
        if (in_array($builder, ['\Illuminate\Database\Eloquent\Builder', 'EloquentBuilder'])) {
            return;
        }
```
This slipped my diligence when I reviewed it, but having `EloquentBuilder` here seems like a hack to me, because I can name the import of the `Builder` anything I want. In fact, when the `Builder` is imported, it does not return here.

Here's an easy case how to reproduce it:
- go to `\Barryvdh\LaravelIdeHelper\Tests\Console\ModelsCommand\GeneratePhpdocWithFqn\Models\Post`
- add the following line to the `use` statements
  ```php
  use Illuminate\Database\Eloquent\Builder;
  ```
- run the test in `\Barryvdh\LaravelIdeHelper\Tests\Console\ModelsCommand\GeneratePhpdocWithFqn\Test::test`
- 💥 with `Failed asserting that 'Exception: array_diff(): Expected parameter 1 to be an array, null given`

Exactly the same problem I see.

---
I don't think adding `Builder` to the `in_array` check is an appropriate fix, because the alias for the import can be anything. Consequently I think having `EloquentBuilder` there is already not correct.

The issue is that this method receives both FQCN and non-FQCN and I've no good proposal how to fix this, except making sure the logic always knows the FQN?

Thank you!

**Comment by KaloyanYosifov** [comment]
Hey @mfn 

Thank you for your time looking into this. I agree with you, that adding the "Builder" into the `in_array` function is going to be a hack. 

Will try to figure out a solution to this and will keep you updated

**Comment by KaloyanYosifov** [comment]
@mfn 

I've added a fix for this in -- https://github.com/barryvdh/laravel-ide-helper/pull/1113

While I was testing I also noticed another bug with this 😅 . If the user has an external eloquent builder he or she will have the new methods in the doc settings, but the based ones like "where, orWhere ...etc" will use the original builder class.

For example:

Let's say we have an external builder to a model "TestModel" called "AwesomeBuilder" and we add a new method to it like "isHavingFun". If the user uses `TestModel::isHavingFun()->where('id', 1)->first()` the autocomplete will work, but if they start with `TestModel::where('id', 1)->isHavingFun()->first()` an error will be shown for `isHavingFun` since the where function returns the original Builder.

---

I have thought of a fix for this but will create a separate pull request since it is not related to this here. 🥂
